File: src/shared/contracts/error_codes.py
"""[SPEC-13A] Business error codes and HTTP/message mappings.

Authority: docs/specs/SPEC-A-contracts.md SPEC-13A "Business Error Codes".

Defines 17 EVID_ error codes grouped by domain:
  EVID_1xxx — Input / project validation
  EVID_2xxx — Gate / workflow
  EVID_3xxx — Agent / LLM
  EVID_4xxx — Artifact
  EVID_5xxx — System

Exports:
  ErrorCode            — Enum of all 17 error codes (values are the string ident).
  ERROR_CODE_HTTP_MAP  — Mapping from code string -> HTTP status (int).
  ERROR_CODE_MESSAGE_MAP — Mapping from code string -> default user message (str).
  lookup_evid(code)    — Return {"code":..., "http_status":..., "message":...}
                         for a given code string, or raise KeyError.
"""

from __future__ import annotations

from enum import Enum
from typing import Dict


class ErrorCode(str, Enum):
    """Business error codes per SPEC-13A."""

    EVID_1001 = "EVID_1001"
    EVID_1002 = "EVID_1002"
    EVID_1003 = "EVID_1003"
    EVID_2001 = "EVID_2001"
    EVID_2002 = "EVID_2002"
    EVID_2003 = "EVID_2003"
    EVID_2004 = "EVID_2004"
    EVID_2005 = "EVID_2005"
    EVID_3001 = "EVID_3001"
    EVID_3002 = "EVID_3002"
    EVID_3003 = "EVID_3003"
    EVID_3004 = "EVID_3004"
    EVID_4001 = "EVID_4001"
    EVID_4002 = "EVID_4002"
    EVID_5001 = "EVID_5001"
    EVID_5002 = "EVID_5002"
    EVID_5003 = "EVID_5003"


ERROR_CODE_HTTP_MAP: Dict[str, int] = {
    ErrorCode.EVID_1001.value: 400,
    ErrorCode.EVID_1002.value: 404,
    ErrorCode.EVID_1003.value: 409,
    ErrorCode.EVID_2001.value: 422,
    ErrorCode.EVID_2002.value: 409,
    ErrorCode.EVID_2003.value: 400,
    ErrorCode.EVID_2004.value: 400,
    ErrorCode.EVID_2005.value: 409,
    ErrorCode.EVID_3001.value: 500,
    ErrorCode.EVID_3002.value: 504,
    ErrorCode.EVID_3003.value: 400,
    ErrorCode.EVID_3004.value: 500,
    ErrorCode.EVID_4001.value: 404,
    ErrorCode.EVID_4002.value: 409,
    ErrorCode.EVID_5001.value: 503,
    ErrorCode.EVID_5002.value: 503,
    ErrorCode.EVID_5003.value: 500,
}


ERROR_CODE_MESSAGE_MAP: Dict[str, str] = {
    ErrorCode.EVID_1001.value: "Description too short (min 10 chars)",
    ErrorCode.EVID_1002.value: "Project not found",
    ErrorCode.EVID_1003.value: "Project is not active",
    ErrorCode.EVID_2001.value: "Gate check failed",
    ErrorCode.EVID_2002.value: "Gate check already in progress",
    ErrorCode.EVID_2003.value: "Phase {N} cannot be skipped",
    ErrorCode.EVID_2004.value: "Cannot rollback to phase {N}",
    ErrorCode.EVID_2005.value: "Active tasks exist, cannot advance",
    ErrorCode.EVID_3001.value: "Agent {name} failed: {reason}",
    ErrorCode.EVID_3002.value: "Agent {name} timed out after {N}s",
    ErrorCode.EVID_3003.value: "Task {id} cannot be cancelled (status: {status})",
    ErrorCode.EVID_3004.value: "LLM call failed: {provider} {error}",
    ErrorCode.EVID_4001.value: "Artifact not found for phase {N}",
    ErrorCode.EVID_4002.value: "Artifact damaged: {path}",
    ErrorCode.EVID_5001.value: "System not ready: {check_name} failed",
    ErrorCode.EVID_5002.value: "Worker unavailable",
    ErrorCode.EVID_5003.value: "Database error",
}


def lookup_evid(code: str) -> dict:
    """Look up error code metadata.

    Args:
        code: Error code string, e.g. "EVID_1001".

    Returns:
        dict with keys "code", "http_status", "message".

    Raises:
        KeyError: If the code is not known.
    """
    if code not in ERROR_CODE_HTTP_MAP:
        raise KeyError(f"Unknown error code: {code}")
    return {
        "code": code,
        "http_status": ERROR_CODE_HTTP_MAP[code],
        "message": ERROR_CODE_MESSAGE_MAP[code],
    }


__all__ = [
    "ErrorCode",
    "ERROR_CODE_HTTP_MAP",
    "ERROR_CODE_MESSAGE_MAP",
    "lookup_evid",
]

File: src/shared/contracts/error_codes.ts
// [SPEC-13A] Business error codes — TypeScript mirror of
// src/shared/contracts/error_codes.py.
// Authority: docs/specs/SPEC-A-contracts.md SPEC-13A "Business Error Codes".
// 17 EVID_ error codes with HTTP status and default message mappings.

export const ERROR_CODES = {
  EVID_1001: "EVID_1001",
  EVID_1002: "EVID_1002",
  EVID_1003: "EVID_1003",
  EVID_2001: "EVID_2001",
  EVID_2002: "EVID_2002",
  EVID_2003: "EVID_2003",
  EVID_2004: "EVID_2004",
  EVID_2005: "EVID_2005",
  EVID_3001: "EVID_3001",
  EVID_3002: "EVID_3002",
  EVID_3003: "EVID_3003",
  EVID_3004: "EVID_3004",
  EVID_4001: "EVID_4001",
  EVID_4002: "EVID_4002",
  EVID_5001: "EVID_5001",
  EVID_5002: "EVID_5002",
  EVID_5003: "EVID_5003",
} as const;

export type ErrorCodeType = (typeof ERROR_CODES)[keyof typeof ERROR_CODES];

/** HTTP status code for each EVID error code. */
export const ERROR_CODE_HTTP_MAP: Record<ErrorCodeType, number> = {
  [ERROR_CODES.EVID_1001]: 400,
  [ERROR_CODES.EVID_1002]: 404,
  [ERROR_CODES.EVID_1003]: 409,
  [ERROR_CODES.EVID_2001]: 422,
  [ERROR_CODES.EVID_2002]: 409,
  [ERROR_CODES.EVID_2003]: 400,
  [ERROR_CODES.EVID_2004]: 400,
  [ERROR_CODES.EVID_2005]: 409,
  [ERROR_CODES.EVID_3001]: 500,
  [ERROR_CODES.EVID_3002]: 504,
  [ERROR_CODES.EVID_3003]: 400,
  [ERROR_CODES.EVID_3004]: 500,
  [ERROR_CODES.EVID_4001]: 404,
  [ERROR_CODES.EVID_4002]: 409,
  [ERROR_CODES.EVID_5001]: 503,
  [ERROR_CODES.EVID_5002]: 503,
  [ERROR_CODES.EVID_5003]: 500,
};

/** Default user-facing message for each EVID error code. */
export const ERROR_CODE_MESSAGE_MAP: Record<ErrorCodeType, string> = {
  [ERROR_CODES.EVID_1001]: "Description too short (min 10 chars)",
  [ERROR_CODES.EVID_1002]: "Project not found",
  [ERROR_CODES.EVID_1003]: "Project is not active",
  [ERROR_CODES.EVID_2001]: "Gate check failed",
  [ERROR_CODES.EVID_2002]: "Gate check already in progress",
  [ERROR_CODES.EVID_2003]: "Phase {N} cannot be skipped",
  [ERROR_CODES.EVID_2004]: "Cannot rollback to phase {N}",
  [ERROR_CODES.EVID_2005]: "Active tasks exist, cannot advance",
  [ERROR_CODES.EVID_3001]: "Agent {name} failed: {reason}",
  [ERROR_CODES.EVID_3002]: "Agent {name} timed out after {N}s",
  [ERROR_CODES.EVID_3003]: "Task {id} cannot be cancelled (status: {status})",
  [ERROR_CODES.EVID_3004]: "LLM call failed: {provider} {error}",
  [ERROR_CODES.EVID_4001]: "Artifact not found for phase {N}",
  [ERROR_CODES.EVID_4002]: "Artifact damaged: {path}",
  [ERROR_CODES.EVID_5001]: "System not ready: {check_name} failed",
  [ERROR_CODES.EVID_5002]: "Worker unavailable",
  [ERROR_CODES.EVID_5003]: "Database error",
};

export interface ErrorCodeInfo {
  code: ErrorCodeType;
  http_status: number;
  message: string;
}

/**
 * Look up error code metadata.
 * Throws if the code string is not a known EVID code.
 */
export function lookupEvid(code: string): ErrorCodeInfo {
  const httpStatus = ERROR_CODE_HTTP_MAP[code as ErrorCodeType];
  const message = ERROR_CODE_MESSAGE_MAP[code as ErrorCodeType];
  if (httpStatus === undefined || message === undefined) {
    throw new Error(`Unknown error code: ${code}`);
  }
  return { code: code as ErrorCodeType, http_status: httpStatus, message };
}
